File: tcpdump/analyze_traffic/js/analyze_traffic_page.js
// Copyright 2021 BlueCat Networks (USA) Inc. and its affiliates. All Rights Reserved.
$(document).ready(function () {
    getConfigurations();
    checkEnableButton();
});


function getConfigurations() {
    $.ajax({
        url: `/api/v1/analyze_traffic/configurations`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            var result = [];
            result.push(`<option value='Please Select'>Please Select</option>`)
            for (i = 0; i < data.length; i++) {
                result.push(`<option value='${data[i]['name']}'>${data[i]['name']}</option>`)
            }
            $('#configuration').html(result);
            getServers();

        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}


function getServers() {
    var configName = $('#configuration').val();
    $('#server').html(`<option value='Please Select'>Please Select</option>
                       <option value='${BAM_NAME}(${BAM_IP})'>${BAM_NAME}(${BAM_IP})</option>`)
    if (configName.includes('Please Select')) {
        $('#server').val(BAM_NAME + '(' + BAM_IP + ')');
        getInterface();
        return false
    } else {
        $('#interfaceName').val("Please Select")
    }

    $.ajax({
        url: `/api/v1/analyze_traffic/configurations/${configName}/servers`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            var result = [];
            for (i = 0; i < data.length; i++) {
                serverIp = getProperty(data[i], 'defaultInterfaceAddress')
                result.push(`<option value='${data[i]['name']}(${serverIp})'>${data[i]['name']}(${serverIp})</option>`)
            }
            $('#server').html($('#server').html() + result);
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}


function checkEnableButton() {
    var interfaceName = $("#interfaceName").val();
    var packageToCapture = $("#packageToCapture").val();
    var maxCaptureFile = $("#maxCaptureFile").val();
    var maxCaptureTime = $("#maxCaptureTime").val().trim();
    disabledButton();

    if ((interfaceName != "Please Select" && packageToCapture && maxCaptureTime) || (interfaceName != "Please Select" && maxCaptureFile && maxCaptureTime)){
        enabledButton();
    }
}

function getInterface() {
    var configName = $('#configuration').val();
    if (configName.includes('Please Select')){
        configName = '';
    }
    var server = $('#server').val();
    if (server.includes('Please Select')){
        return
    }
    server_ip = server.substring(server.indexOf('(') + 1,server.indexOf(')'))
    $.ajax({
        url: `/api/v1/analyze_traffic/interfaces?configuration_name=${configName}&server_ip=${server_ip}`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
                var result = [];
                result.push(`<option value='Please Select'>Please Select</option>`)
                for (i = 0; i < data.length; i++) {
                    result.push(`<option value='${data[i]}'>${data[i]}</option>`)
                }
                $('#interfaceName').html(result);
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}

function startTCPDump() {
    var configName = $('#configuration').val();
    if (configName.includes('Please Select')){
        configName = '';
    }
    var server = $('#server').val();
    server_ip = server.substring(server.indexOf('(') + 1,server.indexOf(')'));
    var interface = $('#interfaceName').val();
    var port = $('#port').val().trim();
    if (port == "") {
        port = $('#port').val(null)
    }
    var packetsToCapture = $('#packageToCapture').val().trim();
    if (packetsToCapture == "") {
        packetsToCapture = $('#packageToCapture').val(null)
    }
    var optional = $('#Optional').val().trim();
    var maxCaptureFile = $('#maxCaptureFile').val().trim();
    if (maxCaptureFile == "") {
        maxCaptureFile = $('#maxCaptureFile').val(null)
    }
    var maxCaptureTime = $('#maxCaptureTime').val().trim();
    var data = {
        "configuration_name": configName,
        "server_ip": server_ip,
        "interface": interface,
        "port": Number(port),
        "packets_to_capture": Number(packetsToCapture),
        "max_capture_file": Number(maxCaptureFile),
        "max_capture_time": Number(maxCaptureTime),
        "optional": optional
    }
    $.ajax({
        url: `/api/v1/analyze_traffic/start`,
        type: 'post',
        data: JSON.stringify(data),
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            definedSuccessMessage(data)
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}

function stopTCPDump() {
    var configName = $('#configuration').val();
    if (configName.includes('Please Select')){
        configName = '';
    }
    var server = $('#server').val();
    server_ip = server.substring(server.indexOf('(') + 1,server.indexOf(')'));
    var data = {
        "configuration_name": configName,
        "server_ip": server_ip
    }
    $.ajax({
        url: `/api/v1/analyze_traffic/stop`,
        type: 'post',
        data: JSON.stringify(data),
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            definedSuccessMessage(data)
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}

function downloadCapture() {
    var configName = $('#configuration').val();
    if (configName.includes('Please Select')){
        configName = '';
    }
    var server = $('#server').val();
    server_ip = server.substring(server.indexOf('(') + 1,server.indexOf(')'));
    var $content = DEFAULT_CONTENT;
    $('body').prepend($content);
    var href = window.location.origin + '/api/v1/analyze_traffic/download?' + `configuration_name=${configName}&server_ip=${server_ip}`
    window.location.href = href;
    setTimeout(
        function () {
            $("#wait").remove();
        }, 3500);
}

function checkStatus() {
    var configName = $('#configuration').val();
    if (configName.includes('Please Select')){
        configName = '';
    }
    var server = $('#server').val();
    server_ip = server.substring(server.indexOf('(') + 1,server.indexOf(')'));
    $.ajax({
        url: `/api/v1/analyze_traffic/status?configuration_name=${configName}&server_ip=${server_ip}`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            if ('Running'.includes(data) || 'Stopped'.includes(data)){
                definedSuccessMessage(`Tcpdump status is ${data}`)
            }
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}
    
function getProperty(entity, name) {
    var result = "";
    var properties = entity['properties'].split("|");
    for (p = 0; p < properties.length; p++) {
        if (properties[p].includes(`${name}`)) {
            result = properties[p].split("=")[1];
        }
    }
    return result
}

function validateInputField() {
    var port = $('#port').val().trim();
    if (port < 0 || port > 65535){
        definedSmallErrorMessage('errorPort', "Invalid Port. It's must be integer, between 0 and 65535.");
        disabledButton();
    } else {
        $('#errorPort').html('');
    }
    var packetsToCapture = $('#packageToCapture').val().trim();
    if (packetsToCapture == ''){
        $('#errorPackageToCapture').html('');
    } else if (packetsToCapture <= 0 || packetsToCapture >= 1000){
        definedSmallErrorMessage('errorPackageToCapture', "Invalid Packets To Capture. It's must be integer, between 1 and 999.");
        disabledButton();
    } else {
        $('#errorPackageToCapture').html('');
    }
    var maxCaptureFile = $('#maxCaptureFile').val().trim();
    if (maxCaptureFile == ''){
        $('#errorMaxCaptureFile').html('');
    } else if (maxCaptureFile <= 0 || maxCaptureFile >= 20){
        definedSmallErrorMessage('errorMaxCaptureFile', "Invalid Max Capture File Size. It's must be integer, between 1 and 19.");
        disabledButton();
    } else {
        $('#errorMaxCaptureFile').html('');
    }
    var maxCaptureTime = $('#maxCaptureTime').val().trim();
    if (maxCaptureTime == ''){
        $('#errorMaxCaptureTime').html('');
    } else if (maxCaptureTime <= 0 || maxCaptureTime >= 300){
        definedSmallErrorMessage('errorMaxCaptureTime', "Invalid Max Capture Time. It's must be integer, between 1 and 299.");
        disabledButton();
    } else {
        $('#errorMaxCaptureTime').html('');
    }
}

function disabledButton(){
    $("#start").prop("disabled", true);
    $("#stop").prop("disabled", true);
    $("#download").prop("disabled", true);
    $("#checkStatus").prop("disabled", true);
}

function enabledButton(){
    $("#start").prop("disabled", false);
    $("#stop").prop("disabled", false);
    $("#download").prop("disabled", false);
    $("#checkStatus").prop("disabled", false);
}

function clearField(){
    $("#interfaceName").val('Please Select');
    $("#port").val('');
    $("#packageToCapture").val('');
    $("#maxCaptureFile").val('');
    $("#maxCaptureTime").val('');
}

File: tcpdump/analyze_traffic/analyze_traffic_page.py
# Copyright 2021 BlueCat Networks (USA) Inc. and its affiliates. All Rights Reserved.
import os
import sys
import traceback

from flask import render_template, g, make_response, jsonify, request, send_file
from flask_restplus import Resource
from bluecat import route, util
import config.default_config as config
from main_app import app, api
from ...common.exceptions import UserException, InvalidParam
from ...common.tcpdump import TcpDump
from ...common.common import get_bam_name, prepare_cmd_tcpdump, validate_configuration_and_server
from ...common.constants import TCP_CAPTURE_FILE_NAME
from .libs.models import analyze_traffic_models
from .libs.parsers import analyze_traffic_parsers


def module_path():
    encoding = sys.getfilesystemencoding()
    return os.path.dirname(os.path.abspath(__file__))


analyze_traffic_ns = api.namespace(
    'analyze_traffic',
    path='/analyze_traffic',
    description='Analyze Traffic TcpDump by Gateway',
)


@route(app, '/analyze_traffic/analyze_traffic_endpoint')
@util.workflow_permission_required('analyze_traffic_page')
@util.exception_catcher
@util.ui_secure_endpoint
def analysis_analysis_page():
    return render_template(
        'analyze_traffic_page.html',
        text=util.get_text(module_path(), config.language),
        options=g.user.get_options(),
        bam_name=get_bam_name(),
        bam_ip=g.user.get_api_netloc()
    )


@analyze_traffic_ns.route('/start')
class StartedTcpDump(Resource):
    @util.workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @analyze_traffic_ns.expect(analyze_traffic_models.started_analyze_traffic_model)
    def post(self):
        """ Start tcpdump"""
        try:
            tcpdump = None
            data = request.get_json()
            host = validate_configuration_and_server(data.get('configuration_name'), data.get('server_ip'))
            # Retrieve interfaces network from host
            tcpdump = TcpDump(host=host)
            interfaces = tcpdump.get_interface_network()
            tcpdump.disconnect()

            # Prepare the commandline to run Tcpdump
            keypair_cmd = prepare_cmd_tcpdump(data, interfaces)
            tcpdump = TcpDump(host=host)
            response = tcpdump.start(keypair_cmd)
            return jsonify(response), 201

        except InvalidParam as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as ex:
            if tcpdump:
                tcpdump.disconnect()
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 500)


@analyze_traffic_ns.route('/stop')
class StoppedTcpDump(Resource):
    @util.workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @analyze_traffic_ns.expect(analyze_traffic_models.stopped_analyze_traffic_model)
    def post(self):
        """ Stop tcpdump"""
        try:
            tcpdump = None
            data = request.get_json()
            host = validate_configuration_and_server(data.get('configuration_name'), data.get('server_ip'))
            tcpdump = TcpDump(host=host)
            response = tcpdump.stop()
            tcpdump.disconnect()
            return jsonify(response), 201

        except InvalidParam as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as ex:
            if tcpdump:
                tcpdump.disconnect()
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 500)


@analyze_traffic_ns.route('/status')
class TcpDumpStatus(Resource):
    @util.workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @analyze_traffic_ns.expect(analyze_traffic_parsers.parser)
    def get(self):
        """ Get tcpdump status"""
        try:
            tcpdump = None
            server_ip = request.args.get('server_ip')
            config_name = request.args.get('configuration_name')
            host = validate_configuration_and_server(config_name, server_ip)
            tcpdump = TcpDump(host=host)
            response = tcpdump.get_status()
            tcpdump.disconnect()
            return jsonify(response), 200

        except InvalidParam as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as e:
            if tcpdump:
                tcpdump.disconnect()
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@analyze_traffic_ns.route('/configurations')
class ConfigurationCollection(Resource):
    @util.rest_workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self):
        """ Get all known configuration(s). """
        try:
            configurations = g.user.get_api().get_configurations()
            result = [config_entity.to_json() for config_entity in configurations]
            return jsonify(result), 200
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@analyze_traffic_ns.route('/configurations/<string:configuration>')
class Configuration(Resource):
    @util.rest_workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self, configuration):
        """ Get configuration with specified name. """
        try:
            configuration = g.user.get_api().get_configuration(configuration)
            result = configuration.to_json()
            return jsonify(result), 200
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@analyze_traffic_ns.route('/configurations/<string:configuration>/servers')
class ServersList(Resource):
    @util.workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self, configuration):
        """ Get all known Server(s) with specified Configuration name. """
        try:
            configuration = g.user.get_api().get_configuration(configuration)
            servers = configuration.get_servers()
            servers = [server.to_json() for server in servers]
            return jsonify(servers), 200
        except UserException as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 500)


@analyze_traffic_ns.route('/download')
class Download(Resource):
    @util.workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @analyze_traffic_ns.expect(analyze_traffic_parsers.parser)
    def get(self):
        """ Support download capture.cap file from BAM and return local path"""
        try:
            tcpdump = None
            server_ip = request.args.get('server_ip')
            config_name = request.args.get('configuration_name')
            host = validate_configuration_and_server(config_name, server_ip)
            tcpdump = TcpDump(host=host)
            local_capture_path = tcpdump.download_and_get_local_capture_file()
            tcpdump.disconnect()
            return send_file(local_capture_path,
                             as_attachment=True,
                             attachment_filename=TCP_CAPTURE_FILE_NAME), 200

        except InvalidParam as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as e:
            if tcpdump:
                tcpdump.disconnect()
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@analyze_traffic_ns.route('/interfaces')
class InterfacesList(Resource):
    @util.workflow_permission_required('analyze_traffic_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @analyze_traffic_ns.expect(analyze_traffic_parsers.parser)
    def get(self):
        """ Get all Interface(s) of server with specified server ip. """
        try:
            tcpdump = None
            server_ip = request.args.get('server_ip')
            config_name = request.args.get('configuration_name')
            host = validate_configuration_and_server(config_name, server_ip)
            tcpdump = TcpDump(host=host)
            interface_list = tcpdump.get_interface_network()
            tcpdump.disconnect()
            return jsonify(interface_list), 200

        except InvalidParam as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as ex:
            if tcpdump:
                tcpdump.disconnect()
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 500)